### ISSUE TITLE
Error compiling on macOS: "linking with `rust-lld` failed: signal: 6"

### ISSUE THREAD

**Issue by gabrielbitencourt** [phil-opp/blog_os#832]
`blog_os $ (post-02 *) cargo xbuild --verbose`
```
+ "rustc" "--print" "sysroot"
+ "rustc" "--print" "target-list"
+ "/Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/bin/cargo" "rustc" "-p" "alloc" "--release" "--manifest-path" "/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/Cargo.toml" "--target" "/Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json" "-v" "--" "-Z" "force-unstable-if-unmarked"
   Compiling core v0.0.0 (/Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore)
   Compiling compiler_builtins v0.1.32
     Running `rustc --crate-name build_script_build /Users/gabrielbitencourt/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.32/build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 -Cembed-bitcode=no --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="mem"' -C metadata=06d061cde57c546e -C extra-filename=-06d061cde57c546e --out-dir /var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/build/compiler_builtins-06d061cde57c546e -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/deps --cap-lints allow`
     Running `rustc --crate-name core --edition=2018 /Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -Cembed-bitcode=no -C metadata=7304f5bc416492bd -C extra-filename=-7304f5bc416492bd --out-dir /var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps --target /Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/deps -Cembed-bitcode=yes`
     Running `/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/build/compiler_builtins-06d061cde57c546e/build-script-build`
   Compiling rustc-std-workspace-core v1.99.0 (/Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/tools/rustc-std-workspace-core)
     Running `rustc --crate-name rustc_std_workspace_core --edition=2018 /Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/tools/rustc-std-workspace-core/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -Cembed-bitcode=no -C metadata=ce8f8a2c7ffbd193 -C extra-filename=-ce8f8a2c7ffbd193 --out-dir /var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps --target /Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/deps --extern core=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps/libcore-7304f5bc416492bd.rmeta -Cembed-bitcode=yes`
     Running `rustc --crate-name compiler_builtins /Users/gabrielbitencourt/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.32/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -Cembed-bitcode=no --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="mem"' -C metadata=27a913a5962eaf62 -C extra-filename=-27a913a5962eaf62 --out-dir /var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps --target /Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/deps --extern core=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps/librustc_std_workspace_core-ce8f8a2c7ffbd193.rmeta --cap-lints allow -Cembed-bitcode=yes --cfg 'feature="unstable"'`
   Compiling alloc v0.0.0 (/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP)
     Running `rustc --crate-name alloc --edition=2018 /Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src/liballoc/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -Cembed-bitcode=no -Z force-unstable-if-unmarked -C metadata=44f7ec07754415f9 -C extra-filename=-44f7ec07754415f9 --out-dir /var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps --target /Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps -L dependency=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/release/deps --extern compiler_builtins=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps/libcompiler_builtins-27a913a5962eaf62.rmeta --extern core=/var/folders/59/fkj059tj5wj32mz_z0jw8z4m0000gn/T/cargo-xbuild.e6Cmycr2MsiP/target/x86_64-blog_os/release/deps/libcore-7304f5bc416492bd.rmeta -Cembed-bitcode=yes`
    Finished release [optimized] target(s) in 29.92s
+ RUSTFLAGS="--sysroot /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot"
+ RUSTDOCFLAGS="--sysroot /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot"
+ "/Users/gabrielbitencourt/.rustup/toolchains/nightly-x86_64-apple-darwin/bin/cargo" "build" "--verbose"
       Fresh bootloader v0.9.3
   Compiling blog_os v0.1.0 (/Users/gabrielbitencourt/Projetos/rust/blog_os)
     Running `rustc --crate-name blog_os --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -Cembed-bitcode=no -C debuginfo=2 -C metadata=dfa431f3382434f6 -C extra-filename=-dfa431f3382434f6 --out-dir /Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps --target /Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json -C incremental=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/incremental -L dependency=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps -L dependency=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/debug/deps --extern bootloader=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/libbootloader-d882c97cdd9071e4.rlib --sysroot /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot`
error: linking with `rust-lld` failed: signal: 6
  |
  = note: "rust-lld" "-flavor" "gnu" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.147igtv00nnhwzpj.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.1i3zeq1udzk766gf.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.2e3rjc0fduahisxd.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.2hmkptsnwtudsitc.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.2x49vdcdzla9tzn3.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.3m1utom4y46zi9hk.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.3mhn9pk65gyf1l76.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.50plrxwhexnisb1p.rcgu.o" "-o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6" "--gc-sections" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/debug/deps" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib" "-Bstatic" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib/librustc_std_workspace_core-ce8f8a2c7ffbd193.rlib" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib/libcore-7304f5bc416492bd.rlib" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib/libcompiler_builtins-27a913a5962eaf62.rlib" "-Bdynamic"
  = note: dyld: lazy symbol binding failed: Symbol not found: __ZN4llvm8InitLLVMC1ERiRPPKcb
            Referenced from: /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-apple-darwin/bin/rust-lld
            Expected in: /Users/gabrielbitencourt/.rustup/toolchains/stable-x86_64-apple-darwin/lib/libLLVM.dylib
          
          dyld: Symbol not found: __ZN4llvm8InitLLVMC1ERiRPPKcb
            Referenced from: /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-apple-darwin/bin/rust-lld
            Expected in: /Users/gabrielbitencourt/.rustup/toolchains/stable-x86_64-apple-darwin/lib/libLLVM.dylib
          
          

error: aborting due to previous error

error: could not compile `blog_os`.

Caused by:
  process didn't exit successfully: `rustc --crate-name blog_os --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -Cembed-bitcode=no -C debuginfo=2 -C metadata=dfa431f3382434f6 -C extra-filename=-dfa431f3382434f6 --out-dir /Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps --target /Users/gabrielbitencourt/Projetos/rust/blog_os/x86_64-blog_os.json -C incremental=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/incremental -L dependency=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps -L dependency=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/debug/deps --extern bootloader=/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/libbootloader-d882c97cdd9071e4.rlib --sysroot /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot` (exit code: 1)
```

**Comment by phil-opp** [comment]
I can't reproduce this error and I never encountered it before. I just reran our CI job for the `post-02` branch using the latest nightly and it also succeeded without problems (including the macOS job): https://github.com/phil-opp/blog_os/runs/816003719?check_suite_focus=true

Could you try running `cargo clean` and try to compile it again?

**Comment by gabrielbitencourt** [comment]
Same error. One thing that caught my eye is

```
error: linking with `rust-lld` failed: signal: 6
  |
  = note: "rust-lld" "-flavor" "gnu" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.147igtv00nnhwzpj.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.1i3zeq1udzk766gf.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.2e3rjc0fduahisxd.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.2hmkptsnwtudsitc.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.2x49vdcdzla9tzn3.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.3m1utom4y46zi9hk.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.3mhn9pk65gyf1l76.rcgu.o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6.50plrxwhexnisb1p.rcgu.o" "-o" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps/blog_os-dfa431f3382434f6" "--gc-sections" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/x86_64-blog_os/debug/deps" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/debug/deps" "-L" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib" "-Bstatic" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib/librustc_std_workspace_core-ce8f8a2c7ffbd193.rlib" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib/libcore-7304f5bc416492bd.rlib" "/Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-blog_os/lib/libcompiler_builtins-27a913a5962eaf62.rlib" "-Bdynamic"
  = note: dyld: lazy symbol binding failed: Symbol not found: __ZN4llvm8InitLLVMC1ERiRPPKcb
            Referenced from: /Users/gabrielbitencourt/Projetos/rust/blog_os/target/sysroot/lib/rustlib/x86_64-apple-darwin/bin/rust-lld
            Expected in: /Users/gabrielbitencourt/.rustup/toolchains/stable-x86_64-apple-darwin/lib/libLLVM.dylib
```

that apparently it was expecting this symbol in the stable version. The stable is my default toolchain, but I've overridden this project to use nightly using `rustup override add nightly` and the `rustup --version` returns `rustc 1.46.0-nightly`. Is there anything that I'm missing?

**Comment by phil-opp** [comment]
Good catch! Sounds like the problem is not in our project, but in your installed Rust toolchain. Apparently the `rust-lld` executable requires a `libLLVM.dylib` library, which is not there somehow. This could mean that something went wrong on your last `rustup update`. I don't know why it looks for it in the stable directory, but I would recommend reinstalling the nightly Rust toolchain (`rustup toolchain remove nightly`  and then `rustup toolchain install nightly`).

**Comment by gabrielbitencourt** [comment]
Yeah, that didn't work either. Even reinstalling rust and setting nightly as my default toolchain didn't work. Only way I got to compile is to backup libLLVM.dylib from stable and replace it with the one from nightly manually, then it works. But that's the ugliest solution ever. Anyway, I'll close this issue since it's probably on cargo-xbuild or rust-lld idk

**Comment by azjkjensen** [comment]
fwiw I ran into a similar issue with libnettle.6.dylib and fixed it by upgrading qemu from 5.0.0 to 5.0.0_2.

```
➜  os git:(master) ✗ qemu-system-x86_64 -drive format=raw,file=target/x86_64-rsos/debug/bootimage-rsos.bin
dyld: Library not loaded: /usr/local/opt/nettle/lib/libnettle.6.dylib
  Referenced from: /usr/local/bin/qemu-system-x86_64
  Reason: image not found
[1]    80337 abort      qemu-system-x86_64 -drive
```
